packages/api/lambdas/cleanExecutions.js

```diff
@@ -1,48 +1,248 @@
+  payloadTimeout
+) => moment().subtract(payloadTimeout, 'days').toDate();
 
-  completeDisable = JSON.parse(completeDisable);
-  nonCompleteDisable = JSON.parse(nonCompleteDisable);
+/**
+ * Clean up Elasticsearch executions that have expired
+ *
+ * @param {number} payloadTimeout - Maximum number of days a record should be held onto
+ * @param {boolean} cleanupRunning - Enable removal of running execution
+ *   payloads
+ * @param {boolean} cleanupNonRunning - Enable removal of execution payloads for
+ *   statuses other than 'running'
+ * @param {number} updateLimit - maximum number of records to update
+ * @param {string} index - Elasticsearch index to cleanup
+ * @returns {Promise<void>}
+*/
+const cleanupExpiredESExecutionPayloads = async (
+  payloadTimeout,
+  cleanupRunning,
+  cleanupNonRunning,
+  updateLimit,
+  index
+) => {
+  const _expiration = getExpirationDate(payloadTimeout);
+  const expiration = _expiration.getTime();
 
-  if (completeDisable && nonCompleteDisable) {
-    return [];
+  const must = [
+    { range: { updatedAt: { lte: expiration } } },
+    {
+      bool: {
+        should: [
+          { exists: { field: 'finalPayload' } },
+          { exists: { field: 'originalPayload' } },
+        ],
+      },
+    },
+  ];
+  const mustNot = [];
+
+  if (cleanupRunning && !cleanupNonRunning) {
+    must.push({ term: { status: 'running' } });
+  } else if (!cleanupRunning && cleanupNonRunning) {
+    mustNot.push({ term: { status: 'running' } });
   }
+  const removePayloadScript = "ctx._source.remove('finalPayload'); ctx._source.remove('originalPayload')";
 
-  const nonCompleteTimeout = Number.parseInt(process.env.nonCompleteExecutionPayloadTimeout, 10);
-  const completeTimeout = Number.parseInt(process.env.completeExecutionPayloadTimeout, 10);
-
-  const configuration = [{
-    name: 'nonCompleteExecutionPayloadTimeout',
-    value: nonCompleteTimeout,
-    originalValue: process.env.nonCompleteExecutionPayloadTimeout,
-  },
-  {
-    name: 'completeExecutionPayloadTimeout',
-    value: completeTimeout,
-    originalValue: process.env.completeExecutionPayloadTimeout,
-  }];
-
-  configuration.forEach((timeout) => {
-    if (!Number.isInteger(timeout.value)) {
-      throw new TypeError(`Invalid number of days specified in configuration for ${timeout.name}: ${timeout.originalValue}`);
-    }
+  const script = { inline: removePayloadScript };
+  const body = {
+    query: {
+      bool: {
+        must,
+        mustNot,
+      },
+    },
+    script: script,
+  };
+  const esClient = await getEsClient();
+  const updateTask = await esClient._client.updateByQuery({
+    index,
+    type: 'execution',
+    size: updateLimit,
+    body,
+    conflicts: 'proceed',
+    wait_for_completion: false,
   });
+  let taskStatus;
+  // this async and poll method allows us to avoid http timeouts
+  // and persist in case of lambda timeout
+  log.info(`launched async ES task id ${updateTask.body.task}`);
+  do {
+    sleep(10000);
```

Comment:
Are we waiting here for the client to become available? Is there a better way to do this other than a hard-coded arbitrary time?

Reply:
Are we really ok with this possibly taking longer than the Lambda timeout? It'd make debugging difficult (we'd see it start but no updates after) and is there anything _after_ this we expect/need to execute?

Reply:
the es cluster is running an asynchronous task while we wait and poll to see if it's done
in the comment just above you can see a mention of avoiding http timeouts and persist if lambda times out. but this means we can't just 'await' it

Reply:
I'd rather it take longer than lambda timeout than crash when lambda times out? but there is a log on line 93 which gives you the task ID to follow up and check on a running job if the lambda crashes independently

Reply:
Hmm so dumb question: if we were to make the PG and ES deletes truly idempotent, we wouldn't have to sleep or await? Could we fire off the ES process here and just return the task id every time rather than a success/failure response? Basically making this operate like our async operations (because it is 😄 ).

Reply:
so:
PG can't do that, I was able to find a module https://github.com/tembo-io/pg_later which seems to try to do this, and maybe that's something worth coming back to
ES can do that, it seemed like the best approach would be to attempt to do it *synchronously* with logs in the lambda logs and a job following it but with the understanding that *actually doing the cleanup* was the priority. especially since we want to do it in relatively small batches rather than potentially slam the ES cluster with a mega job of cleaning up X billion executions
that was my thought process, but yes ES is basically doing exactly this deep down

Reply:
are we satisfied with just ES launching as an off brand async operation with lambda printing a task ID and then attempting to follow it?

Reply:
Sounds like yes. Do we still want to sleep in that case?

Reply:
I think we do, I'd like our lambda function to attempt to follow through on it without requiring the user to check on it